### PR TITLE
Extend http.ServerResponse + Header Building

### PR DESCRIPTION
**Pull request by asmahood**
### Description of the PR
This PR extends `http.ServerResponse` to add helper methods for ease of use. `APIServer` and `http.IncomingMessage` are accessible on the extended class. Also adds automatic response header building through `new APIServer(serverOptions, listener)`. This PR only includes support for a fair amount of headers, but not all of them. Headers are built once on server startup. Also adds routes in test/server.js for the helper methods.

### Areas That Need Work
- [x] Finished
~~- `ResponseNextra#sendFile`~~
~~- Currently uses `fs.readFileSync` because it was the only method that didn't error out. Will happily change in favour of a better solution~~

- [x] Finished
~~- `headerBuilder.js`~~
~~- Needs in-depth testing to see if all headers set properly~~

### Changes Proposed in this Pull Request (List new items in CHANGELOG.MD)
- Extend http.ServerResponse
- Add many helper methods
- Add test routes
- Add automatic header building

### Semver Classification
- [ ] This PR only includes documentation or non-code changes.
- [ ] This PR fixes a bug and does not change the (intended) framework interface.
- [x] This PR adds methods or properties to the framework interface.
- [ ] This PR removes or renames methods or properties in the framework interface.

### CODE REVIEW

File: src/APIServer.js [hunk 1]
```diff
@@ -1,19 +1,89 @@
...
 class APIServer extends Server {
 
-	constructor(...args) {
-		super(...args);
+	/**
+	 * Creates a new HTTP server
+	 * @param {Object} options Any object
+	 * @param {Function} listener A listener for HTTP requests. Arguments are `request` and `response`
+	 */
+	constructor(options = {}, listener) {
```
Comment (kyranet) [inline src/APIServer.js]: Options should be last argument.

File: src/ResponseNextra.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	 * @param {string} path A route on the server
+	 */
+	redirect(path) {
+		this.writeHead(303, { Location: `http://${this.request.headers.host}${path}` });
+		super.end();
+	}
+
+}
+
+const MIMETYPES = {
```
Comment (kyranet) [inline src/ResponseNextra.js]: Should be exported as `ResponseNextra.MIMETYPES`. Possibly move it to constants and allow the developer modify them to add more values.

File: test/server.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 const { APIServer } = require('../index');
 
-const server = new APIServer(async (request, response) => {
+const server = new APIServer(undefined, async (request, response) => {
```
Comment (kyranet) [inline test/server.js]: Remove the `undefined` with the first requested change.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: src/ResponseNextra.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		 * @type {string}
+		 */
+		this.contentType = null;
+	}
+
+	/**
+	 * Sets response headers and sends data to the body
+	 * @param {(string|Buffer)} data Any data to be sent
+	 */
+	end(data) {
```
Comment (MrJacz) [inline src/ResponseNextra.js]: Should use `...args` considering end takes 1, 2 or 3 arguments

**Comment by asmahood** [comment]
This PR should be pretty much finished, in terms of code changes. Only things that may need changing is better error checking in options.headers building. This PR also now relies on PR (#1) to properly send files. `ResponseNextra#sendFile` may be toxic based on it randomly working and not working during testing.

**Comment by asmahood** [comment]
Now it's finished 😄

**Comment by vladfrangu** [comment]
I'll be home tommorow so expect #1 to be done and merge-ready ~unless kyra does the edits cause I have allow maintainers to edit enabled~

**Comment by asmahood** [comment]
Alright, I have made all the changes I want to make. This PR should be good to go, and header options should be error proof, but you never know! 👍

**Pull request merged**